File: ThreeIntegerSum/three_integer_sum.py
#!/usr/bin/env python3

class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        sums = []
        print(nums)
        for i, target in enumerate(nums):
            if i > 0 and nums[i-1] == target:
                continue
            l, r = i+1, len(nums) - 1
            while l < r:
                partial_sum = nums[l] + nums[r]
                if partial_sum > -target:
                    r -= 1

                elif partial_sum < -target:
                    l += 1

                else:
                    sums.append([target, nums[l], nums[r]])
                    l += 1

                    while nums[l] == nums[l-1] and l < r:
                        l += 1
        return sums

File: ThreeIntegerSum/three_integer_sum.cpp
#include <algorithm>
#include <iostream>
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        std::sort(nums.begin(), nums.end());
        vector<vector<int>> triplets;
        for (int i = 0; i < nums.size()-2; ++i){
            if (0 < i && nums[i] == nums[i-1]){
                continue;}

            int l = i+1, r = nums.size()-1;
            std::cout << nums[i] << std::endl;
            while (l < r){
                if (-nums[i] < nums[l] + nums[r])
                    --r;
                else if (nums[l] + nums[r] < -nums[i])
                    ++l;
                else{
                    triplets.push_back({nums[l], nums[i], nums[r]});
                    ++l;
                    --r;
                    while (l < r && nums[l] == nums[l-1])
                        ++l;
                }
            }
        }
        return triplets;
    }
};

